File: app/util/JsonDeserializer.java
/*
 * Copyright (c) 2017 Exam Consortium
 *
 * Licensed under the EUPL, Version 1.1 or - as soon they will be approved by the European Commission - subsequent
 * versions of the EUPL (the "Licence");
 * You may not use this work except in compliance with the Licence.
 * You may obtain a copy of the Licence at:
 *
 * https://joinup.ec.europa.eu/software/page/eupl/licence-eupl
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the Licence is distributed
 * on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Licence for the specific language governing permissions and limitations under the Licence.
 */

package util;

import com.fasterxml.jackson.databind.JsonNode;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonDeserializationContext;
import com.google.gson.JsonElement;
import org.joda.time.DateTime;
import org.joda.time.format.ISODateTimeFormat;
import play.Logger;

import java.lang.reflect.Type;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Date;

public final class JsonDeserializer {

    private JsonDeserializer() {
    }

    private static final GsonBuilder gsonBuilder = new GsonBuilder();
    static {
        gsonBuilder.registerTypeAdapter(Date.class, new DateDeserializer());
        gsonBuilder.registerTypeAdapter(DateTime.class, new DateTimeDeserializer());
    }
    private static final Gson gson = gsonBuilder.create();

    private static class DateDeserializer implements com.google.gson.JsonDeserializer<Date> {

        @Override
        public Date deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
            try {
                return new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").parse(json.getAsString());
            } catch (ParseException e) {
                try {
                    return new Date(json.getAsLong());
                } catch (RuntimeException e2) {
                    Logger.warn("Failed to parse date " + json.getAsString());
                }
            }
            return null;
        }
    }

    private static class DateTimeDeserializer implements com.google.gson.JsonDeserializer<DateTime> {

        @Override
        public DateTime deserialize(JsonElement json, Type typeOfT, JsonDeserializationContext context) {
            try {
                return ISODateTimeFormat.dateTime().parseDateTime(json.getAsString());
            } catch (IllegalArgumentException e) {
                try {
                    return new DateTime(json.getAsLong());
                } catch (RuntimeException e2) {
                    Logger.warn("Failed to parse date " + json.getAsString());
                }
            }
            return null;
        }
    }

    public static <T> T deserialize(Class<T> model, JsonNode node) {
        return gson.fromJson(node.toString(), model);
    }
}

File: app/services/json/JsonDeserializer.scala
// SPDX-FileCopyrightText: 2024 The members of the EXAM Consortium
//
// SPDX-License-Identifier: EUPL-1.2

package services.json

import com.fasterxml.jackson.databind.JsonNode
import com.google.gson.GsonBuilder
import com.google.gson.JsonDeserializationContext
import com.google.gson.JsonDeserializer
import com.google.gson.JsonElement
import com.google.gson.JsonPrimitive
import com.google.gson.JsonSerializationContext
import com.google.gson.JsonSerializer
import play.api.Logging

import java.lang.reflect.Type
import java.text.SimpleDateFormat
import java.time.format.DateTimeFormatter
import java.time.{Instant, LocalTime}
import java.util.Date
import scala.util.{Success, Try}

object JsonDeserializer:
  private val builder = new GsonBuilder()
  builder.registerTypeAdapter(classOf[Date], new DateDeserializer())
  builder.registerTypeAdapter(classOf[Instant], new InstantTypeAdapter())
  builder.registerTypeAdapter(classOf[LocalTime], new LocalTimeTypeAdapter())
  private val gson = builder.create()

  def deserialize[T](model: Class[T], node: JsonNode): T = gson.fromJson(node.toString, model)

class DateDeserializer extends JsonDeserializer[Date] with Logging:
  override def deserialize(
      json: JsonElement,
      typeOfT: Type,
      context: JsonDeserializationContext
  ): Date =
    Try(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'").parse(json.getAsString)) match
      case Success(d) => d
      case _ =>
        Try(new Date(json.getAsLong)) match
          case Success(dt) => dt
          case _ =>
            Try(new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSSXXX").parse(json.getAsString)) match
              case Success(dt) => dt
              case _ =>
                logger.warn(s"Failed to parse date ${json.getAsString}")
                null

class InstantTypeAdapter extends JsonSerializer[Instant] with JsonDeserializer[Instant]
    with Logging:
  override def serialize(
      src: Instant,
      typeOfSrc: Type,
      context: JsonSerializationContext
  ): JsonElement =
    new JsonPrimitive(DateTimeFormatter.ISO_INSTANT.format(src))

  override def deserialize(
      json: JsonElement,
      typeOfT: Type,
      context: JsonDeserializationContext
  ): Instant =
    Try(Instant.parse(json.getAsString)) match
      case Success(i) => i
      case _ =>
        Try(Instant.ofEpochMilli(json.getAsLong)) match
          case Success(i) => i
          case _ =>
            logger.warn(s"Failed to parse instant ${json.getAsString}")
            null

class LocalTimeTypeAdapter extends JsonSerializer[LocalTime] with JsonDeserializer[LocalTime]
    with Logging:
  override def serialize(
      src: LocalTime,
      typeOfSrc: Type,
      context: JsonSerializationContext
  ): JsonElement =
    new JsonPrimitive(src.format(DateTimeFormatter.ISO_LOCAL_TIME))

  override def deserialize(
      json: JsonElement,
      typeOfT: Type,
      context: JsonDeserializationContext
  ): LocalTime =
    Try(LocalTime.parse(json.getAsString, DateTimeFormatter.ISO_LOCAL_TIME)) match
      case Success(t) => t
      case _ =>
        logger.warn(s"Failed to parse local time ${json.getAsString}")
        null
